### PR TITLE
ssh: refactor auth logic to use updated api

### PR DESCRIPTION
**Pull request by kralicky**
## Summary

Several of the internal SSH auth api messages were changed in https://github.com/pomerium/envoy-custom/pull/257. This PR updates the pomerium-side auth logic to use the updated messages. 

The most important change here is the way it handles tracking the state of the auth flow. Previously there were two messages in our api named `PublicKeyAuthMethodResponse` and `KeyboardInteractiveAuthMethodResponse` which were not related to the actual SSH protocol messages but were intended to hold public key state and oauth state, respectively. These messages were replaced with a single `AuthContext` message which holds the combined state. The `Auth` type was refactored to use the `AuthContext` for state tracking instead of the two other messages. 

This lets us (mostly) decouple the protocol auth methods from the parts of the auth context that are obtained during each method. For example a successful public key auth flow will add public key information to the auth context, but in some cases it can also add session info to the auth context as well, like if there is an existing valid session bound to that public key (and that session is still authorized). 

In addition, the keyboard-interactive method is currently used to facilitate the oauth flow, but we can also use it to implement new auth patterns such as two-person rule, or other features we might want to add in the future that would make use of interactive messages or user input.

This PR also contains improvements to the existing integration test suite and lots of new integration tests which cover more of the auth code. The unit tests in auth_test.go cover similar areas of the code but those tests mock out the evaluator responses and databroker client (mostly to test internal failures or other edge cases) so they are not sufficient for testing the whole auth flow like the integration tests are.

Note that there are a few locations with placeholder code (auth.go:114, auth.go:363-375); this is for the upcoming two-person rule feature, since support for that auth flow was used to gauge what refactors would be needed. In that example the keyboard-interactive flow can be used for two different purposes, either to obtain session context or (if there is already session context, and another flag is set) to trigger the two person flow. This does rely on using the same `HandleKeyboardInteractiveMethodRequest` for both, and inspecting the current state to decide what to do next, which is maybe not perfect but I didn't want to make too many changes yet to the `AuthInterface` outside of splitting up the stream info/auth info args. In the future it may be useful to combine the two methods into one, but there wasn't an immediately obvious way to do so.

## Related issues

https://github.com/pomerium/envoy-custom/pull/257

## User Explanation

None

## AI disclosure

None

## Checklist

- [x] reference any related issues
- [x] updated unit tests
- [x] add appropriate label (`enhancement`, `bug`, `breaking`, `dependencies`, `ci`)
- [x] disclosed AI usage (or wrote "none") per AI_POLICY.md
- [x] ready for review

### CODE REVIEW

File: config/policy.go [hunk 1]
```diff
@@ -823,6 +823,10 @@ func (p *Policy) Validate() error {
 			return fmt.Errorf("config: mcp authorization_server_url must be https, got %q", u.Scheme)
 		}
 	}
+
+	if p.IsSSH() && p.AllowPublicUnauthenticatedAccess {
```
Comment (kenjenkins) [inline config/policy.go:834]: This looks like a change in behavior. I think this PR would be easier to review if we split off any behavior changes to a separate PR.

File: pkg/ssh/stream.go [hunk 2]
```diff
@@ -45,25 +50,66 @@ type KeyboardInteractiveQuerier interface {
...
 }
 
-type (
-	PublicKeyAuthMethodResponse           = AuthMethodResponse[extensions_ssh.PublicKeyAllowResponse]
-	KeyboardInteractiveAuthMethodResponse = AuthMethodResponse[extensions_ssh.KeyboardInteractiveAllowResponse]
-)
+func (r *AuthMethodResponse) Validate() {
+	if r.AllowMethod {
+		if r.NextRequiredAuthMethod == "" && !r.NoFurtherMethodsRequired {
+			panic("invalid AuthMethodResponse: either NextRequiredAuthMethod or NoFurtherMethodsRequired must be set if AllowMethod is true")
```
Comment (kenjenkins) [inline pkg/ssh/stream.go:63]: Ordinarily I would expect a Validate() method to return an error rather than panicking. Is there a reason than panic() is better here?

File: pkg/ssh/ssh_int_test.go
Comment (kenjenkins) [inline pkg/ssh/ssh_int_test.go:1]: Would it better to split off the integration test changes into a separate PR as well? If we merge any test improvements first I think that would help to demonstrate that the refactoring changes are sound.

Comment (kralicky) [reply pkg/ssh/ssh_int_test.go:1]: Yes that is a good idea.

Comment (kralicky) [reply pkg/ssh/ssh_int_test.go:1]: https://github.com/pomerium/pomerium/pull/6732

File: pkg/ssh/stream.go (hunk 2 above)
Comment (kralicky) [reply pkg/ssh/stream.go:63]: This is a failsafe to try to catch future programmer error. If any of these panics are hit, it means there is either a serious bug in the auth logic, or something was changed in a way that breaks an existing invariant and additional code changes are required.

File: config/policy.go (hunk 1 above)
Comment (kralicky) [reply config/policy.go:834]: This didn't work before, it would function the same as allow_any_authenticated_user, which was technically a bug. After these changes it still won't work, but the user will be denied instead and an error will be logged. A policy with this option will never be usable so this just gives a better error message for this case. There are still other ways to construct an invalid policy but this is easy to validate compared to arbitrary ppl.

File: pkg/ssh/auth.go [hunk 3]
```diff
@@ -108,6 +111,7 @@ func NewAuth(
 	currentConfig *atomic.Pointer[config.Config],
 	tracerProvider oteltrace.TracerProvider,
 	codeIssuer code.Issuer,
+	_ any, // temporary placeholder
```
Comment (kenjenkins) [inline pkg/ssh/auth.go:114]: What is this for?

File: pkg/ssh/auth.go [hunk 4]
```diff
@@ -479,12 +604,21 @@ func (a *Auth) reportLoginCodeFailure(
 
 var errAccessDenied = status.Error(codes.PermissionDenied, "access denied")
 
-func (a *Auth) EvaluateDelayed(ctx context.Context, info StreamAuthInfo, user api.UserRequest) error {
-	req, err := a.sshRequestFromStreamAuthInfo(ctx, info, user)
+func (a *Auth) EvaluateDelayed(ctx context.Context, streamInfo StreamInfo, authInfo StreamAuthInfo, user api.UserRequest) error {
+	if user.Hostname() == "" {
+		panic("bug: EvaluateDelayed called with empty hostname")
```
Comment (kenjenkins) [inline pkg/ssh/auth.go:609]: It looks like a panic here will cause Pomerium to exit. Can we return an error instead?

File: pkg/ssh/auth.go (hunk 3 above)
Comment (kralicky) [reply pkg/ssh/auth.go:114]: It's replaced here https://github.com/pomerium/pomerium/pull/6736/changes#diff-dfa698e1898e30a60acd64f3a596c2bdb7e42b257fa046d2f6382f03185e0f9c in a follow up. Changes to the number of arguments in this method affect mock usage and iirc was causing some annoying conflicts since there are a few other prs that depend on this one. It may or may not be still important. This code was split from the above linked pr at some point.

File: pkg/ssh/auth.go (hunk 4 above)
Comment (kralicky) [reply pkg/ssh/auth.go:609]: No, this is an invariant check. It should be impossible to hit unless the api is used incorrectly.

File: pkg/ssh/stream.go [hunk 5]
```diff
@@ -253,7 +287,12 @@ func (sh *StreamHandler) OnClusterHealthUpdate(_ context.Context, event *datav3.
 }
 
 func (sh *StreamHandler) Terminate(err error) {
```
Comment (kenjenkins) [inline pkg/ssh/stream.go:290]: Is it possible for Terminate() to be called multiple times? Do we need any check here in case it was already called?

**Review by kenjenkins — APPROVED** [review]

File: pkg/ssh/stream.go (hunk 5 above)
Comment (kralicky) [reply pkg/ssh/stream.go:290]: Hmm, maybe. If the context was already canceled then canceling it again would be a no-op, but then terminateErr might end up being something different than the cancel cause. But the only time terminateErr is ever checked is in a situation where terminateFunc is nil. Perhaps that is too confusing.

**Review by alexandreLamarre — APPROVED** [review]
Some TODOs that look leftover. Might be addressed in subsequent PRs, but worth taking a second look at before merging
